lapack/testlapack/matgen.go

```diff
@@ -738,3 +738,505 @@ func applyReflector(qh blas64.General, q blas64.General, v []float64) {
 		}
 	}
 }
+
+func dlattb(kind int, uplo blas.Uplo, trans blas.Transpose, n, kd int, ab []float64, ldab int, rnd *rand.Rand) (diag blas.Diag, b []float64) {
+	switch {
+	case kind < 1 || 18 < kind:
+		panic("bad matrix kind")
+	case (6 <= kind && kind <= 9) || kind == 17:
+		diag = blas.Unit
+	default:
+		diag = blas.NonUnit
+	}
+
+	if n == 0 {
+		return
+	}
+
+	const (
+		unfl   = dlamchS
+		ulp    = dlamchE * dlamchB
+		smlnum = unfl
+		bignum = (1 - ulp) / smlnum
+
+		eps   = dlamchP
+		small = 0.25 * (dlamchS / eps)
+		large = 1 / small
+		badc2 = 0.1 / eps
+	)
+	badc1 := math.Sqrt(badc2)
+
+	var cndnum float64
+	switch {
+	case kind == 2 || kind == 8:
+		cndnum = badc1
+	case kind == 3 || kind == 9:
+		cndnum = badc2
+	default:
+		cndnum = 2
+	}
+
+	uniformM11 := func() float64 {
+		return 2*rnd.Float64() - 1
+	}
+
+	// Allocate the right-hand side and fill it with random numbers.
+	// The pathological matrix types below overwrite it with their
+	// custom vector.
+	b = make([]float64, n)
+	for i := range b {
+		b[i] = uniformM11()
+	}
+
+	bi := blas64.Implementation()
+	switch kind {
+	default:
+		panic("test matrix type not implemented")
+
+	case 1, 2, 3, 4, 5:
+		// Non-unit triangular matrix
+		// TODO(vladimir-ch)
+		var kl, ku int
+		switch uplo {
+		case blas.Upper:
+			ku = kd
+			kl = 0
+			// IOFF = 1 + MAX( 0, KD-N+1 )
+			// PACKIT = 'Q'
+			// 'Q' => store the upper triangle in band storage scheme
+			//        (only if matrix symmetric or upper triangular)
+		case blas.Lower:
+			ku = 0
+			kl = kd
+			// IOFF = 1
+			// PACKIT = 'B'
+			// 'B' => store the lower triangle in band storage scheme
+			//        (only if matrix symmetric or lower triangular)
+		}
+		anorm := 1.0
+		switch kind {
+		case 4:
+			anorm = small
+		case 5:
+			anorm = large
+		}
+		_, _, _ = kl, ku, anorm
+		// // DIST = 'S' // UNIFORM(-1, 1)
+		// // MODE = 3   // MODE = 3 sets D(I)=CNDNUM**(-(I-1)/(N-1))
+		// // TYPE = 'N' // If TYPE='N', the generated matrix is nonsymmetric
+		// CALL DLATMS( N, N, DIST, ISEED, TYPE, B, MODE, CNDNUM, ANORM,
+		// $            KL, KU, PACKIT, AB( IOFF, 1 ), LDAB, WORK, INFO )
+		panic("test matrix type not implemented")
+
+	case 6:
+		// Matrix is the identity.
+		if uplo == blas.Upper {
+			for i := 0; i < n; i++ {
+				// Fill the diagonal with random numbers.
+				ab[i*ldab] = rnd.NormFloat64()
```

Comment:
Is there a reason to use random rather than the systematic non-unitary elements in the original?

Reply:
Not really. Maybe I was not consistent in what I put on the diagonal of Unit matrices, I'll check again.

Reply:
Done.